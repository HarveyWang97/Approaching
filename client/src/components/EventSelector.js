import React, { Component } from 'react';
import '../css/ItemSelector.css';
import Icon from './popup/Icon';
import { connect } from 'react-redux';
import  * as actions from '../actions';
import MultiSelect from "@kenshooui/react-multi-select";

class EventSelector extends ComponentÂ {
    componentDidMount(){
        let arr = [];
        for(var i=0;i<this.props.events.length;i++){
            arr.push({label:this.props.events[i].name,id:this.props.events[i]._id});
        }
        this.setState({items:arr});
        this.setState({selectedItems:this.props.payload.formatted_details});
    }

    constructor(props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.state = {
          items: [],
          selectedItems: []
        };
      }
    
    handleChange(selectedItems) {
        this.setState({ selectedItems });
    }

    handleSubmit() {
        this.props.payload.handleSubmit(this.state.selectedItems);
        this.props.toggleEventSelector();
    }

    render() {
        const { items, selectedItems } = this.state;
        console.log("items", selectedItems);
        return (
          <div className="item-selector">
          <div className='item-selector-inner'>
                    <div className='close-icon' onClick={() => this.props.toggleEventSelector()}>
                        <Icon  iconName='times'  />
                    </div>
                    <div className='submit' onClick={this.handleSubmit.bind(this)}>
                        <Icon iconName='save'/>
                    </div>
                    <div className='caption' >
                        Please select from following events:
                    </div>
                    <div style={{marginTop:'70px'}}>
                        {this.state.items.length > 0 ?
                         <MultiSelect
                            items={items}
                            selectedItems={selectedItems}
                            onChange={this.handleChange}
                        />: <h1>no items</h1>
                        } 
                        
                    </div>
                </div>
        
          
          </div>
        );
    }


}

function mapStateToProps(state){
    return {
        rawItems:state.items['rawItems'],
        events: state.events.rawEvents
    }
}

export default connect(mapStateToProps, actions)(EventSelector);
