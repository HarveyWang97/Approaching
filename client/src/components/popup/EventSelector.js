import React, { Component } from 'react';
import '../../assets/styles/ItemSelector.css';
import Icon from '../common/Icon';
import { connect } from 'react-redux';
import  * as actions from '../../actions';
import MultiSelect from "@kenshooui/react-multi-select";

/**
 * Class representing the EventSelector section in Item list
 * @class
 */
class EventSelector extends ComponentÂ {
    /**
     * @function
     * @param 
     * @return {void}
     * This function helps to change the events to the form of [{id:xxx,label:xxx}...]
     */

    componentDidMount(){
        let arr = [];
        if(this.props.rawItems !== null && this.props.rawItems !== undefined) {
            for(var i=0;i<this.props.events.length;i++){
                arr.push({label:this.props.events[i].name,id:this.props.events[i]._id});
            }
        }
        this.setState({items:arr});
        this.setState({selectedItems:this.props.payload.formatted_details});
    }

    constructor(props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.state = {
            items: [],
            selectedItems: []
        };
    }

    /**
     * This function helps to submit the changes in selectedItems
     * @param {Array<item>} selectedItems 
     * @return {void}
     */
    handleChange(selectedItems) {
        this.setState({ selectedItems });
    }


    /**
     * This function helps to submit the selected items and close the pop up window
     * @return {void} 
     */
    handleSubmit() {
        this.props.payload.handleSubmit(this.state.selectedItems);
        this.props.toggleEventSelector();
    }

    /**
     * Render the eventSelector html block.
     * @returns {html} Returns an html object of eventSelector.
     */
    render() {
        const { items, selectedItems } = this.state;
        console.log("items", selectedItems);
        return (
            <div className="item-selector">
                <div className='item-selector-inner'>
                    <div className='close-icon' onClick={() => this.props.toggleEventSelector()}>
                        <Icon  iconName='times'  />
                    </div>
                    <div className='submit' onClick={this.handleSubmit.bind(this)}>
                        <Icon iconName='save'/>
                    </div>
                    <div className='caption' >
                        Please select from following events:
                    </div>
                    <div className='multi-select-container'>
                        {this.state.items.length > 0 ?
                         <MultiSelect
                            items={items}
                            selectedItems={selectedItems}
                            onChange={this.handleChange}
                            responsiveHeight='100%' 
                        />: <h1>no items</h1>
                        }
                    </div>
                </div>
            </div>
        );
    }
}

function mapStateToProps(state){
    return {
        rawItems:state.items['rawItems'],
        events: state.events.rawEvents
    }
}

export default connect(mapStateToProps, actions)(EventSelector);
