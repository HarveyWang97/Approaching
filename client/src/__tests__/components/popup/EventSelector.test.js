import React from 'react';
import mockStore from '../../../__testutils__//mockStore';
import {shallow} from 'enzyme';
import EventSelector from '../../../components/popup/EventSelector';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const init_state = {
    events:{
        rawEvents:[
            {_id:'abc',name:'hulu onsite'},
            {_id:'df',name:'facebook onsite'},
            {_id:'cd',name:'apple interview'}
        ]
    },
    items:{
        rawItems:[]
    }
};

const payload = {formatted_details:[]};

describe("test ItemSelector",() => {
    const wrapper = shallow(<EventSelector payload={payload}/>,
        {
            context:{store:mockStore(init_state)}   
        });
    
    it("render as expected",() => {
        expect(wrapper).toMatchSnapshot();
    });

    it("has the items set correctly",() => {
        expect(wrapper.dive().state('items')).toEqual([
            {id:'abc',label:'hulu onsite'},
            {id:'df',label:'facebook onsite'},
            {id:'cd',label:'apple interview'}
        ]);
    });
    
})