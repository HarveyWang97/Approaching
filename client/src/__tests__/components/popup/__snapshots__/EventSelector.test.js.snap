// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ItemSelector render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(EventSelector)
    payload={
      Object {
        "formatted_details": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "deleteEvent": [Function],
      "deleteItem": [Function],
      "events": Array [
        Object {
          "_id": "abc",
          "name": "hulu onsite",
        },
        Object {
          "_id": "df",
          "name": "facebook onsite",
        },
        Object {
          "_id": "cd",
          "name": "apple interview",
        },
      ],
      "fetchEvents": [Function],
      "fetchItems": [Function],
      "fetchProfile": [Function],
      "fetchUser": [Function],
      "insertEvent": [Function],
      "insertItem": [Function],
      "insertUser": [Function],
      "payload": Object {
        "formatted_details": Array [],
      },
      "rawItems": Array [],
      "toggleEventSelector": [Function],
      "toggleItemSelector": [Function],
      "togglePictureEditor": [Function],
      "togglePopup": [Function],
      "updateEmail": [Function],
      "updateEvent": [Function],
      "updateItem": [Function],
      "updateNotifyTime": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "deleteEvent": [Function],
        "deleteItem": [Function],
        "events": Array [
          Object {
            "_id": "abc",
            "name": "hulu onsite",
          },
          Object {
            "_id": "df",
            "name": "facebook onsite",
          },
          Object {
            "_id": "cd",
            "name": "apple interview",
          },
        ],
        "fetchEvents": [Function],
        "fetchItems": [Function],
        "fetchProfile": [Function],
        "fetchUser": [Function],
        "insertEvent": [Function],
        "insertItem": [Function],
        "insertUser": [Function],
        "payload": Object {
          "formatted_details": Array [],
        },
        "rawItems": Array [],
        "toggleEventSelector": [Function],
        "toggleItemSelector": [Function],
        "togglePictureEditor": [Function],
        "togglePopup": [Function],
        "updateEmail": [Function],
        "updateEvent": [Function],
        "updateItem": [Function],
        "updateNotifyTime": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
